test/utils/FunctionCallGenerator.sol

```diff
@@ -0,0 +1,177 @@
+
+    uint256 public constant FUNCTION_REGISTER = 0;
+    uint256 public constant FUNCTION_REVOKE = 1;
+    uint256 public constant FUNCTION_UPDATE = 2;
+    uint256 public constant TRANSFER_TOKEN = 3;
+    uint256 public constant FUNCTION_COUNT = 4;
+
+    uint256 public constant MAX_DEPTH = 5;
+
+    address private immutable _tokenA;
+    address private immutable _tokenB;
+
+    EnumerableSetLib.Bytes32Set internal _revokedKeyHashes;
+
+    constructor(address tokenA, address tokenB) {
+        _tokenA = tokenA;
+        _tokenB = tokenB;
+    }
+
+    function _randomTestKey() internal returns (TestKey memory newKey) {
+        newKey = TestKeyManager.withSeed(KeyType.Secp256k1, vm.randomUint());
+    }
+
+    function _registerCall(TestKey memory newKey) internal virtual returns (HandlerCall memory) {
+        return CallUtils.initHandlerDefault().withCall(CallUtils.encodeRegisterCall(newKey)).withCallback(
+            abi.encodeWithSelector(IHandlerGhostCallbacks.ghost_RegisterCallback.selector, newKey.toKey())
+        );
+    }
+
+    function _revokeCall(bytes32 keyHash) internal virtual returns (HandlerCall memory) {
+        return CallUtils.initHandlerDefault().withCall(CallUtils.encodeRevokeCall(keyHash)).withCallback(
+            abi.encodeWithSelector(IHandlerGhostCallbacks.ghost_RevokeCallback.selector, keyHash)
+        );
+    }
+
+    function _updateCall(bytes32 keyHash, Settings settings) internal virtual returns (HandlerCall memory) {
+        return CallUtils.initHandlerDefault().withCall(CallUtils.encodeUpdateCall(keyHash, settings)).withCallback(
+            abi.encodeWithSelector(IHandlerGhostCallbacks.ghost_UpdateCallback.selector, keyHash)
+        );
+    }
+
+    function _tokenTransferCall(address token, address to, uint256 amount)
+        internal
+        virtual
+        returns (HandlerCall memory)
+    {
+        return CallUtils.initHandlerDefault().withCall(
+            CallUtils.initDefault().withTo(token).withData(
+                abi.encodeWithSelector(ERC20.transfer.selector, to, amount)
+            )
+        );
+    }
+
+    // Ghost keys are persisted after the callbacks are triggered
+    function _getRandomGhostKeyHash(uint256 seed) internal view returns (bytes32) {
+        if (_ghostKeyHashes.values().length == 0) {
+            return bytes32(0);
+        }
+        return _ghostKeyHashes.values()[seed % _ghostKeyHashes.values().length];
+    }
+
+    function _getSettingsForKeyHash(bytes32 keyHash) internal virtual returns (Settings) {}
+
+    /**
+     * @notice Generate a random function call with equal weighting between function types
+     * @param randomSeed Random seed for generation
+     * @param depth Current recursion depth
+     * @return A call object for the generated function
+     */
+    function generateRandomFunctionHandlerCall(uint256 randomSeed, uint256 depth) public returns (HandlerCall memory) {
+        vm.assume(randomSeed < type(uint256).max);
```

Comment:
Why do you need to add this line?

Reply:
bc the recursion below does seed + 1 for the recursive call

Reply:
seems like this can be removed now?

Reply:
yes!